### PR TITLE
Return full canonical product URL within product object

### PR DESCRIPTION
**Pull request by austris-argalis**
### Description

Added canonical URL to graphql schema along with the necessary resolvers.

### Fixed Issues (if relevant)

1. magento/graphql-ce#2: Return full canonical product URL within product object

### Manual testing scenarios

1. Issue CURL request to a running Magento 2.3 store like this:
`curl -X POST -H "Content-Type: application/json" --data '{"query": "{products(filter: {sku: {neq: \"simple1\"}}) {items {id,canonical_url}}}","variables": [],"operationName": null}' http://magento23.local/graphql`
2. Assert that valid product URLs can be found in the response

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/CatalogGraphQl/Test/Integration/Model/ProductsTest.php
Comment (paliarush) [inline app/code/Magento/CatalogGraphQl/Test/Integration/Model/ProductsTest.php]: Please convert this test into web API functional test.

See [\Magento\GraphQl\Catalog\ProductViewTest::testQueryAllFieldsSimpleProduct](https://github.com/magento/magento2/blob/b3372064f574fd5aece0617009ee317db589febe/dev/tests/api-functional/testsuite/Magento/GraphQl/Catalog/ProductViewTest.php#L35) as an example.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Product/CanonicalUrl.php
Comment (paliarush) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Product/CanonicalUrl.php]: All classes and methods should be documented.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Product/CanonicalUrl.php
Comment (paliarush) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Product/CanonicalUrl.php]: Please use `@inheritdoc` instead of duplicating parent annotations.

File: app/code/Magento/CatalogGraphQl/Test/Unit/Model/Resolver/Product/CanonicalUrlTest.php
Comment (paliarush) [inline app/code/Magento/CatalogGraphQl/Test/Unit/Model/Resolver/Product/CanonicalUrlTest.php:18]: FYI, unit tests are optional and should be added only when it makes sense (e.g. when you can test method API without mocks). Since the test is already completed, we can keep it.

**Review by paliarush — CHANGES_REQUESTED** [review]

**Comment by paliarush** [comment]
@austris-argalis Thank you for your contribution, feel free to contact me directly regarding requested changes at magentocommeng.slack.com (send an email to engcom@magento.com to join).

**Comment by austris-argalis** [comment]
@paliarush thanks for the review!
Hope it's better now.

**Comment by magento-engcom-team** [comment]
Hi @austris-argalis 
Please sign the [Contributor License Agreement](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=7) so that we can proceed with the merge.

Thank you

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=7) <br/>All committers have signed the CLA.

**Review by paliarush — APPROVED** [review]
PR looks good and can be merged after CLA is signed by @austris-argalis

**Comment by magento-engcom-team** [comment]
@austris-argalis thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by paliarush** [comment]
@austris-argalis the test failed CICD environment, please check this:

```
Magento\GraphQl\Catalog\ProductViewTest::testQueryAllFieldsSimpleProduct
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'http://magento.url/index.php/catalog/product/view/_ignore_category/1/id/553/s/simple-product/'
+'http://magento.url/index.php/simple-product.html'
```

**Review by DanielRuf — APPROVED** [review]

**Comment by austris-argalis** [comment]
Hello @paliarush 

Thanks for the info!
Tried reproducing the problem but cannot really get the same results no matter what.
Is the CI server running some specific configurations I could try and replicate?

Been running tests on a fresh install of 2.3-develop over php 7.1.16 with mysql 5.7.
image

also running full graphql api testsuite does not fail this particular test:
image

**Comment by paliarush** [comment]
@austris-argalis I cannot reproduce the issue locally either. Some settings must be different on the instance running tests.

With current implementation, the test may still pass, while the public API changes (e.g. if we change URL generation logic).

Let's rewrite the test to compare the URL to predefined static value. Such test would correctly fail if public GraphQL interface changes for some reason. Just take base URL and append "simple-product.html" to it before assertion.

**Comment by misha-kotov** [comment]
Approved :white_check_mark:

**Comment by paliarush** [comment]
@austris-argalis your PR has been merged, thanks for the contribution, you are the first contributor to Magento's GraphQL track!

**Comment by keharper** [comment]
Added canonical_url to the [Products endpoint](https://devdocs.magento.com/guides/v2.3/graphql/reference/products.html) documentation.

**Pull request merged**
